addons/barcodes/static/src/js/barcode_events.js

```diff
@@ -0,0 +1,158 @@
+odoo.define('web.BarcodeEvents', function(require) {
+"use strict";
+
+var core = require('web.core');
+var mixins = core.mixins;
+
+var BarcodeEvents = core.Class.extend(mixins.PropertiesMixin, {
+    timeout: null,
+    key_pressed: {},
+    buffered_key_events: [],
+    // Regexp to match a barcode input and extract its payload
+    // Note: to build in init() if prefix/suffix can be configured
+    regexp: /(.{3,})[\n\r\t]+/,
+    // By knowing the terminal character we can interpret buffered keys
+    // as a barcode as soon as it's encountered (instead of waiting x ms)
+    suffix: /[\n\r\t]+/,
+    // Keys from a barcode scanner are usually processed as quick as possible,
+    // but some scanners can use an intercharacter delay (we support <= 50 ms)
+    max_time_between_keys_in_ms: 55,
+
+    init: function(parent) {
+        mixins.PropertiesMixin.init.call(this);
+        this.setParent(parent);
+        // Keep a reference of the handler functions to use when adding and removing event listeners
+        this.__keydown_handler = this.keydown_handler.bind(this);
+        this.__keyup_handler = this.keyup_handler.bind(this);
+        this.__handler = this.handler.bind(this);
+        // Bind event handler once the DOM is loaded
+        $(this.start.bind(this));
+    },
+
+    handle_buffered_keys: function() {
+        var str = this.buffered_key_events.reduce(function(memo, e) { return memo + String.fromCharCode(e.which) }, '');
+        var match = str.match(this.regexp);
+
+        if (match) {
+            var barcode = match[1];
+
+            core.bus.trigger('barcode_scanned', barcode);
+
+            // Dispatch a barcode_scanned DOM event to elements that have barcode_events="true" set.
+            if (this.buffered_key_events[0].target.getAttribute("barcode_events") === "true")
+                $(this.buffered_key_events[0].target).trigger('barcode_scanned', barcode);
+        } else {
+            this.resend_buffered_keys();
+        }
+
+        this.buffered_key_events = [];
+    },
+
+    resend_buffered_keys: function() {
+        for(var i = 0; i < this.buffered_key_events.length; i++) {
+            var old_event = this.buffered_key_events[i];
```

Comment:
it does not matter for performance, obviously.  The problem is that the scope of that variable is the whole function (funnily enough, not enough people knows this), but it looks like it is local to the loop (it would be if the keyword "let" was used instead, and that we were using ES2015).   So, the variable is actually defined outside of the for loop.  And bad things happens (because we have bad programmers, and we don't use linters)